cli/tests/test_rebase_command.rs

```diff
@@ -631,3 +631,191 @@ fn test_rebase_with_descendants() {
 fn get_log_output(test_env: &TestEnvironment, repo_path: &Path) -> String {
     test_env.jj_cmd_success(repo_path, &["log", "-T", "branches"])
 }
+
+// This behavior illustrates https://github.com/martinvonz/jj/issues/2600
+// The behavior of `rebase -r A -d descendant_of_A` can also be affected or
+// broken depending on how #2600 is fixed, so we test that as well.
+#[test]
+fn test_rebase_with_child_and_descendant_bug_2600() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
+
+    create_commit(&test_env, &repo_path, "base", &[]);
+    create_commit(&test_env, &repo_path, "a", &["base"]);
+    create_commit(&test_env, &repo_path, "b", &["base", "a"]);
+    create_commit(&test_env, &repo_path, "c", &["b"]);
+    let setup_opid = test_env.current_operation_id(&repo_path);
+
+    // Test the setup
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @  c
+    ◉    b
+    ├─╮
+    │ ◉  a
+    ├─╯
+    ◉  base
+    ◉
+    "###);
+
+    let (stdout, stderr) =
+        test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "base", "-d", "root()"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Also rebased 4 descendant commits onto parent of rebased commit
+    Working copy now at: vruxwmqv 1fdab507 c | c
+    Parent commit      : royxmykx 4d413a39 b | b
+    Added 0 files, modified 0 files, removed 1 files
+    "###);
+    // The user would expect unsimplified ancestry here.
+    // ◉  base
+    // │ @  c
+    // │ ◉    b
+    // │ ├─╮
+    // │ │ ◉  a
+    // │ ├─╯
+    // ├─╯
+    // ◉
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    ◉  base
+    │ @  c
+    │ ◉  b
+    │ ◉  a
+    ├─╯
+    ◉
+    "###);
+
+    // This tests the algorithm for rebasing onto descendants. The result should be
+    // simplified if and only if it's simplified in the above case.
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "base", "-d", "b"]);
```

Comment:
So, it turns out that I forgot to `jj op restore` before this and the next test. If I add `jj op restore`-s, they panic. I'll need to debug my algorithm.

The failing commit is https://github.com/ilyagr/jj/commit/rebasetest-fail

Reply:
<details><summary>Here's the failure log</summary>

`````
thread 'test_rebase_with_child_and_descendant_bug_2600' panicked at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5:
Unexpected failure.
code=101
stderr=``````
thread \'main\' panicked at lib/src/dag_walk.rs:113:13:
graph has cycle
stack backtrace:
   0: rust_begin_unwind
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/panicking.rs:72:14
   2: jj_lib::dag_walk::topo_order_forward_ok
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/dag_walk.rs:113:13
   3: jj_lib::dag_walk::topo_order_reverse_ok
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/dag_walk.rs:160:22
   4: jj_lib::dag_walk::topo_order_reverse
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/dag_walk.rs:142:5
   5: jj_lib::rewrite::DescendantRebaser::new
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/rewrite.rs:306:24
   6: jj_lib::repo::MutableRepo::create_descendant_rebaser
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/repo.rs:844:9
   7: jj_lib::repo::MutableRepo::rebase_descendants_return_rebaser
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/repo.rs:861:27
   8: jj_lib::repo::MutableRepo::rebase_descendants_with_options
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/repo.rs:872:12
   9: jj_lib::repo::MutableRepo::rebase_descendants
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/repo.rs:878:9
  10: jj_cli::commands::rebase::rebase_revision
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/commands/rebase.rs:420:22
  11: jj_cli::commands::rebase::cmd_rebase
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/commands/rebase.rs:197:9
  12: jj_cli::commands::run_command
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/commands/mod.rs:187:39
  13: core::ops::function::FnOnce::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
  14: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
  15: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/alloc/src/boxed.rs:2007:9
  16: jj_cli::cli_util::CliRunner::run_internal
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/cli_util.rs:2867:9
  17: jj_cli::cli_util::CliRunner::run
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/cli_util.rs:2884:22
  18: jj::main
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/main.rs:18:5
  19: core::ops::function::FnOnce::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
```
command=`cd "/tmp/jj-test-38KGLy/repo" && env -i HOME="/tmp/jj-test-38KGLy/home" JJ_CONFIG="/tmp/jj-test-38KGLy/config" JJ_EMAIL="test.user@example.com" JJ_OP_HOSTNAME="host.example.com" JJ_OP_TIMESTAMP="2001-02-03T04:05:41+07:00" JJ_OP_USERNAME="test-username" JJ_RANDOMNESS_SEED="35" JJ_TIMESTAMP="2001-02-03T04:05:41+07:00" JJ_USER="Test User" RUST_BACKTRACE="1" SSL_CERT_FILE="/dev/null" "/usr/local/google/home/ilyagr/dev/jj/target/debug/jj" "rebase" "-r" "base" "-d" "b"`
code=101
stdout=""
stderr=```
thread \'main\' panicked at lib/src/dag_walk.rs:113:13:
graph has cycle
stack backtrace:
   0: rust_begin_unwind
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/panicking.rs:72:14
   2: jj_lib::dag_walk::topo_order_forward_ok
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/dag_walk.rs:113:13
   3: jj_lib::dag_walk::topo_order_reverse_ok
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/dag_walk.rs:160:22
   4: jj_lib::dag_walk::topo_order_reverse
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/dag_walk.rs:142:5
   5: jj_lib::rewrite::DescendantRebaser::new
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/rewrite.rs:306:24
   6: jj_lib::repo::MutableRepo::create_descendant_rebaser
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/repo.rs:844:9
   7: jj_lib::repo::MutableRepo::rebase_descendants_return_rebaser
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/repo.rs:861:27
   8: jj_lib::repo::MutableRepo::rebase_descendants_with_options
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/repo.rs:872:12
   9: jj_lib::repo::MutableRepo::rebase_descendants
             at /usr/local/google/home/ilyagr/dev/jj/lib/src/repo.rs:878:9
  10: jj_cli::commands::rebase::rebase_revision
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/commands/rebase.rs:420:22
  11: jj_cli::commands::rebase::cmd_rebase
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/commands/rebase.rs:197:9
  12: jj_cli::commands::run_command
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/commands/mod.rs:187:39
  13: core::ops::function::FnOnce::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
  14: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
  15: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/alloc/src/boxed.rs:2007:9
  16: jj_cli::cli_util::CliRunner::run_internal
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/cli_util.rs:2867:9
  17: jj_cli::cli_util::CliRunner::run
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/cli_util.rs:2884:22
  18: jj::main
             at /usr/local/google/home/ilyagr/dev/jj/cli/src/main.rs:18:5
  19: core::ops::function::FnOnce::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```


stack backtrace:
   0: rust_begin_unwind
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/panicking.rs:72:14
   2: core::panicking::panic_display
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/panicking.rs:193:5
   3: assert_cmd::assert::AssertError::panic::panic_cold_display
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/panic.rs:99:13
   4: assert_cmd::assert::AssertError::panic
             at /usr/local/google/home/ilyagr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/assert_cmd-2.0.12/src/assert.rs:1033:9
   5: core::ops::function::FnOnce::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
   6: core::result::Result<T,E>::unwrap_or_else
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/result.rs:1426:23
   7: assert_cmd::assert::Assert::success
             at /usr/local/google/home/ilyagr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/assert_cmd-2.0.12/src/assert.rs:158:9
   8: test_rebase_command::common::TestEnvironment::jj_cmd_ok
             at ./tests/common/mod.rs:147:22
   9: test_rebase_command::test_rebase_with_child_and_descendant_bug_2600
             at ./tests/test_rebase_command.rs:821:28
  10: test_rebase_command::test_rebase_with_child_and_descendant_bug_2600::{{closure}}
             at ./tests/test_rebase_command.rs:639:52
  11: core::ops::function::FnOnce::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
  12: core::ops::function::FnOnce::call_once
             at /rustc/6cf088810f66fff15d05bf7135c5f5888b7c93b4/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
`````

</details>

Reply:
This happens with or without #2646, by the way. I'll probably record the problem in a bug and some `should_panic` tests, and then think about it later.

Reply:
I will encapsulate bits and pieces first, then perhaps we could consider moving them.